Homework 22

OneAndOnlyAdam opened the pull request.

OneAndOnlyAdam committed:
homework-21

Error for BaseTest line 45 suggested I change it from:
wait = new WebDriverWait (driver, Duration. ofSeconds(10))

To:  wait = (WebDriver) new WebDriverWait(driver,Duration.ofSeconds(10));
Can you tell me why?  When during class the instructor had it like the top example.

In both the BaseTest class there's an issue on lines 60,64,69 and Homework21 lines 26,30,36 with the 'until'. rename or change to get is suggested.  What is wrong with it???

OneAndOnlyAdam committed:
Homework-22

Kept getting variations of this error yesterday:
 org.gradle.api.internal.tasks.testing.TestSuiteExecutionException: Could not complete execution for Gradle Test Executor 9.
The number forllowing Executor changed everytime I ran the test.

Then in class last night Muhammad was getting the same error as well.  He was thinking there may be an issue with Gradle.

Berichicko commented:
. Please finish your homework.

![image](https://github.com/Test-Pro-Bootcamp/A51-QA-Automation-Framework/assets/104761482/880ede9e-5f7d-4f19-bcf9-849b12af72a3)

HINT

![image](https://github.com/Test-Pro-Bootcamp/A51-QA-Automation-Framework/assets/104761482/9d07f15a-ae06-493a-bdaf-08da9a2d64d5)

OneAndOnlyAdam committed:
Homework22

Getting the following error:

org.gradle.api.internal.tasks.testing.TestSuiteExecutionException: Could not complete execution for Gradle Test Executor 1.

Do not know what this error is.

Berichicko commented on src/test/java/PageObjectModel/BasePage.java:
```diff
@@ -0,0 +1,44 @@
+public class BasePage {
+
+
+    WebDriver driver;
+    WebDriverWait wait;
+    Actions actions;
```
I recommend: make the variables protected, because protected access gives the subclass a chance to use the helper method or variable, while preventing a nonrelated class from trying to use it.

```
    protected Actions actions;
    protected WebDriverWait wait;
    protected WebDriver driver;
```

Berichicko commented on src/test/java/PageObjectModel/HomePage.java:
```diff
@@ -0,0 +1,28 @@
+    public HomePage(WebDriver givenDriver) {
+        super(givenDriver);
+    }
+    By firstPlaylist = By.cssSelector(".playlist:nth-child(3)");
+    By playlistNameField = By.cssSelector("[name='name']");
```
You should use encapsulation of non-static and non-final fields.

If a variable is used only by other methods of that class, it should be declared as private.

I recommend:  make all the locators and WebElements private, because private keyword in java allows most restrictive access to variables and methods and offer strongest form of Encapsulation. 

For example

    ```
    private By shuffleBtnLocator = By.cssSelector(".btn-shuffle-all");
    private By firstSongLocator = By.cssSelector(".play");

```

Berichicko commented on src/test/java/PageObjectModel/LoginPage.java:
```diff
@@ -0,0 +1,39 @@
+        super(givenDriver);
+    }
+    //Locators
+    By emailField = By.cssSelector("input[type='email']");
+    By passwordField = By.cssSelector("input[type='password']");
+    By submitBtn = By.cssSelector("button[type='submit']");
```
You should use encapsulation of non-static and non-final fields.

If a variable is used only by other methods of that class, it should be declared as private.

I recommend:  make all the locators and WebElements private, because private keyword in java allows most restrictive access to variables and methods and offer strongest form of Encapsulation. 

For example

    ```
    private By shuffleBtnLocator = By.cssSelector(".btn-shuffle-all");
    private By firstSongLocator = By.cssSelector(".play");

```

Berichicko commented on src/test/java/Homework22.java:
```diff
@@ -0,0 +1,57 @@
+}
+
+
+       /**
+        provideEmail("adam.johnson@testpro.io");
+        providePassword("1Te$t$tudent");
+        clickSubmit();
+        //Right/Context click
+        contextClickPlaylist();
+        //Enter new Playlist name
+        enterNewPlaylistName();
+        //Assertion
+        Assert.assertTrue(doesPlaylistExist());
+    }
+    public void contextClickPlaylist() {
+        WebElement playlistElement = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("#playlists > ul > li:nth-child(3) > a")));
+        actions.contextClick(playlistElement).perform();
+    }
+    public void enterNewPlaylistName(){
+        WebElement playlistInputField = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("[name='name']")));
+        playlistInputField.sendKeys(Keys.chord(Keys.CONTROL,"a", Keys.BACK_SPACE));
+        playlistInputField.sendKeys(newPlaylistName);
+        playlistInputField.sendKeys(Keys.ENTER);
+    }
+    public boolean doesPlaylistExist() {
+        WebElement playlistElement = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//a[(text()='" + newPlaylistName + "']")));
+        return playlistElement.isDisplayed();
+    }
+    }
+**/
```
I recommend not using code comments: for, example, when you have a code that cannot be understood, then comments are not the solution. Try to rewrite the code or rename fields and other elements in a way that the reader will understand the action that you are doing there.

Berichicko commented:
You got this problem, please correct it. Please finish your homework.

![image](https://github.com/Test-Pro-Bootcamp/A51-QA-Automation-Framework/assets/104761482/5256b9d4-6e11-4c9d-97ab-2e9e661d3956)

HINT

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >

<suite name="TestNG">
    <parameter name="BaseUrl" value="https://qa.koel.app/"/>
    <test name="Example Test" preserve-order="false">
        <classes>
            <class name="Homework22"/>
        </classes>

    </test>
</suite>
```

And

```
Actions actions ;

    @BeforeSuite
    static void setupClass() {
        WebDriverManager.chromedriver().setup();
    }

    @BeforeMethod
    @Parameters({"BaseUrl"})
    public void launchBrowser(String BaseUrl) throws InterruptedException{
        //Added ChromeOptions argument to fix websocket error
        ChromeOptions options = new ChromeOptions();
        options.addArguments("--remote-allow-origins=*");

        //Open Chrome Browser
        driver = new ChromeDriver(options);
       actions = new Actions(driver);
        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(10));
        //driver.get(url);
        driver.manage().window().maximize();
        wait = new WebDriverWait(driver, Duration.ofSeconds(5));
        url = BaseUrl;
        navigateToPage();
    }
```

![image](https://github.com/Test-Pro-Bootcamp/A51-QA-Automation-Framework/assets/104761482/4913a4ca-6ab8-4961-bdcb-64b2cb0dab60)
